gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/PubSubImpl.java

```diff
@@ -445,17 +466,35 @@ public SubscriptionId apply(String compositeSubscription) {
 
   @Override
   public Iterator<ReceivedMessage> pull(String subscription, int maxMessages) {
-    // this should set return_immediately to true
-    return null;
+    return get(pullAsync(subscription, maxMessages));
   }
 
   @Override
-  public Future<Iterator<ReceivedMessage>> pullAsync(String subscription, int maxMessages) {
-    // though this method can set return_immediately to false (as future can be canceled) I
-    // suggest to keep it false so sync could delegate to asyc and use the same options
-    // this method also should use the VTKIT thread-pool to renew ack deadline for non consumed
-    // messages
-    return null;
+  public Future<Iterator<ReceivedMessage>> pullAsync(final String subscription, int maxMessages) {
+    PullRequest request = PullRequest.newBuilder().setReturnImmediately(true)
+        .setSubscription(SubscriberApi.formatSubscriptionName(options().projectId(), subscription))
+        .setMaxMessages(maxMessages)
+        .setReturnImmediately(true)
+        .build();
+    Future<PullResponse> response = rpc.pull(request);
```

Comment:
I could not find any other reasonable way of implementing this while keeping the async nature of the method. We could use another thread to wait for the future to complete (basically faking a listenable future) and add messages to the renewer as soon as they are pulled. The problem with this approach is that we cannot run such a thread on the same executor of the "RPC future" (as the pool's threads could potentially be all occupied by waiting tasks, thus starving the RPC tasks). Having an executor just for this feature seemed too much to me.

This seemed like a reasonable compromise to me.